cpp/src/Slice/Parser.cpp

```diff
@@ -4424,18 +4430,37 @@ Slice::Operation::visit(ParserVisitor* visitor, bool)
 
 Slice::Operation::Operation(const ContainerPtr& container,
                             const string& name,
-                            const TypePtr& returnType,
-                            bool returnIsTagged,
-                            int returnTag,
+                            const TaggedDefListTokPtr& returnValues,
                             Mode mode) :
     SyntaxTreeBase(container->unit()),
     Contained(container, name),
     Container(container->unit()),
-    _returnType(returnType),
-    _returnIsTagged(returnIsTagged),
-    _returnTag(returnTag),
+    _usesOutParameters(false),
+    _hasReturnType(!returnValues->v.empty()),
     _mode(mode)
 {
+    for (const auto& returnValue : returnValues->v)
+    {
+        _unit->checkType(returnValue->type);
+        if (returnValue->isTagged)
+        {
+            // Check for a duplicate tag.
+            for (const auto& other : _returnValues)
+            {
+                if (other->tagged() && other->tag() == returnValue->tag)
+                {
+                    _unit->error("tag for return value `" + returnValue->name + "' is already in use");
+                }
+            }
```

Comment:
Can we use a std::set to keep the used tags

 ```suggestion
    std::set<int> tags;
    for (const auto& returnValue : returnValues->v)
    {
        _unit->checkType(returnValue->type);
        if (returnValue->isTagged && !tags.insert(returnValue->tag).second)
        {
            _unit->error("tag for return value `" + returnValue->name + "' is already in use");
        }
```

Reply:
Yes, I think this is pretty clever!